plugins/ltac2/tac2def.ml

```diff
@@ -0,0 +1,86 @@
+(************************************************************************)
+(*         *   The Coq Proof Assistant / The Coq Development Team       *)
+(*  v      *         Copyright INRIA, CNRS and contributors             *)
+(* <O___,, * (see version control and CREDITS file for authors & dates) *)
+(*   \VV/  **************************************************************)
+(*    //   *    This file is distributed under the terms of the         *)
+(*         *     GNU Lesser General Public License Version 2.1          *)
+(*         *     (see LICENSE file for the text of the license)         *)
+(************************************************************************)
+
+open Proofview
+open Tac2ffi
+
+(* Local shortcuts. *)
+type v = Tac2ffi.valexpr
+type 'r of_v = v -> 'r
+type 'r to_v = 'r -> v
+type 'a with_env = Environ.env -> Evd.evar_map -> 'a
+
+let value : valexpr repr = make_repr Fun.id Fun.id
+
+type (_, _, _) spec =
+  | T : 'r to_v -> (v tactic, 'r tactic, 'r) spec
+  | V : 'r to_v -> (v tactic, 'r, 'r) spec
+  | E : 'r to_v -> (v tactic, 'r with_env, 'r) spec
+  | G : 'r to_v -> (v tactic, Goal.t -> 'r, 'r) spec
+  | F : 'a of_v * ('t, 'f, 'r) spec -> (v -> 't , 'a -> 'f, 'r) spec
+
+let tac : type r. r repr -> (v tactic, r tactic, r) spec = fun r ->
+  T(repr_of r)
+
+let tac': type r. r to_v -> (v tactic, r tactic, r) spec = fun tr ->
+  T(tr)
+
+let ret : type r. r repr -> (v tactic, r, r) spec = fun r ->
+  V(repr_of r)
+
+let ret': type r. r to_v -> (v tactic, r, r) spec = fun tr ->
+  V(tr)
+
+let eret : type r. r repr -> (v tactic, r with_env, r) spec = fun r ->
+  E(repr_of r)
+
+let eret': type r. r to_v -> (v tactic, r with_env, r) spec = fun tr ->
+  E(tr)
+
+let gret : type r. r repr -> (v tactic, Goal.t -> r, r) spec = fun r ->
+  G(repr_of r)
+
+let gret': type r. r to_v -> (v tactic, Goal.t -> r, r) spec = fun tr ->
+  G(tr)
+
+let (@->) : type a t f r. a repr -> (t,f,r) spec -> (v -> t, a -> f, r) spec =
+  fun r ar -> F(repr_to r, ar)
+
+let (@-->): type a t f r. a of_v -> (t,f,r) spec -> (v -> t, a -> f, r) spec =
+  fun of_v ar -> F(of_v, ar)
+
+let rec interp_spec : type v f r. (v,f,r) spec -> f -> v = fun ar f ->
+  let open Proofview.Notations in
+  match ar with
+  | T(tr) -> f >>= fun v -> tclUNIT (tr v)
+  | V(tr) -> tclUNIT (tr f)
+  | E(tr) -> tclENV >>= fun e -> tclEVARMAP >>= fun s -> tclUNIT (tr (f e s))
+  | G(tr) -> Goal.enter_one @@ fun g -> tclUNIT (tr (f g))
+  | F(tr,ar) -> fun v -> interp_spec ar (f (tr v))
+
+let rec arity_of : type v f r. (v,f,r) spec -> v Tac2ffi.arity = fun ar ->
+  match ar with
+  | F(_, T(_)) -> Tac2ffi.arity_one
+  | F(_, V(_)) -> Tac2ffi.arity_one
+  | F(_, E(_)) -> Tac2ffi.arity_one
+  | F(_, G(_)) -> Tac2ffi.arity_one
+  | F(_, ar) -> Tac2ffi.arity_suc (arity_of ar)
```

Comment:
```suggestion
  | F (_, (T _ | V _ | E _ | G _)) -> Tac2ffi.arity_one
  | F (_, ar) -> Tac2ffi.arity_suc (arity_of ar)
```
or possibly
```suggestion
  | F (_, ar) -> begin match ar with 
    | T _ | V _ | E _ | G _ -> Tac2ffi.arity_one
    | F _ -> Tac2ffi.arity_suc (arity_of ar)
    end
```

Reply:
Neither of these work, since the patterns are not uniform in the value for the type variables.